errors/errors.go

```diff
@@ -67,79 +76,92 @@ func Wrap(err error, context ...map[string]string) *Error {
 // the current authorisation credentials. The user may need authorising, or if authorised,
 // may not be permitted to perform this action
-func Forbidden(message string, context ...map[string]string) *Error {
-	return errorFactory(ErrForbidden, message, context...)
+func Forbidden(code, message string, params map[string]string) *Error {
+	return errorFactory(errCode(ErrForbidden, code), message, params)
 }
 
 // Unauthorized creates a new error indicating that authentication is required,
 // but has either failed or not been provided.
-func Unauthorized(message string, context ...map[string]string) *Error {
-	return errorFactory(ErrUnauthorized, message, context...)
+func Unauthorized(code, message string, params map[string]string) *Error {
+	return errorFactory(errCode(ErrUnauthorized, code), message, params)
 }
 
-// errorConstructor returns a `*Error` with the specified code, message and context. The main work
-// consists of managing the map[string]string's at the end of the arguments list.
-// In practice we only ever pass in two: the first one is public and will be sent to the client,
-// the second one is private and can contain internal information that is useful for debugging
-func errorFactory(code int, message string, context ...map[string]string) *Error {
+// errorConstructor returns a `*Error` with the specified code, message and params.
+// Builds a stack based on the current call stack
+func errorFactory(code string, message string, params map[string]string) *Error {
 	err := &Error{
-		Code:           code,
-		Message:        message,
-		PrivateContext: map[string]string{},
-		PublicContext:  map[string]string{},
+		Code:    ErrUnknown,
+		Message: message,
+		Params:  map[string]string{},
 	}
-	// The first context map is the PublicContext
-	if len(context) > 0 && context[0] != nil {
-		err.PublicContext = context[0]
+	if len(code) > 0 {
+		err.Code = code
 	}
-	// The second context map is the privateContext
-	if len(context) > 1 && context[1] != nil {
-		err.PrivateContext = context[1]
+	if params != nil {
+		err.Params = params
 	}
 
+	// TODO pass in context.Context
+
 	// Build stack and skip first three lines:
 	//  - stack.go BuildStack()
 	//  - errors.go errorFactory()
 	//  - errors.go public constructor method
-	err.Stack = stack.BuildStack(3)
+	err.StackFrames = stack.BuildStack(3)
 
-	// ... ignore all remaining map[string]string
 	return err
 }
+
+func errCode(prefix, code string) string {
+	if code == "" {
+		return prefix
+	}
+	if prefix == "" {
+		return code
+	}
+	return strings.Join([]string{prefix, code}, ".")
+}
+
+// Matches returns whether the string returned from error.Error() contains the given param string. This means you can
+// match the error on different levels e.g. dotted codes `bad_request` or `bad_request.missing_param` or even on the
+// more descriptive message
+func (p *Error) Matches(match string) bool {
+	return strings.Contains(p.Error(), match)
```

Comment:
Good question. I like it being contains because you can then match on the old style message too. For example if you wrap a traditional error, otherwise you'd have to match on `"internal_service: foo bar message"`